bindinfo/handle.go

```diff
@@ -704,14 +705,22 @@ func GenerateBindSQL(ctx context.Context, stmtNode ast.StmtNode, planHint string
 		bindSQL = bindSQL[updateIdx:]
 		return strings.Replace(bindSQL, "UPDATE", fmt.Sprintf("UPDATE /*+ %s*/", planHint), 1)
 	case *ast.SelectStmt:
-		withIdx := strings.Index(bindSQL, "WITH")
-		selectIdx := strings.Index(bindSQL, "SELECT")
-		if withIdx != -1 && withIdx < selectIdx {
-			selectIdx = withIdx
+		var selectIdx int
+		if n.With != nil {
+			var withSb strings.Builder
+			withIdx := strings.Index(bindSQL, "WITH")
+			restoreCtx := format.NewRestoreCtx(format.RestoreStringSingleQuotes|format.RestoreSpacesAroundBinaryOperation|format.RestoreStringWithoutCharset|format.RestoreNameBackQuotes, &withSb)
+			restoreCtx.DefaultDB = defaultDB
+			n.With.Restore(restoreCtx)
+			withEnd := withIdx + len(withSb.String())
+			tmp := strings.Replace(bindSQL[withEnd:], "SELECT", fmt.Sprintf("SELECT /*+ %s*/", planHint), 1)
```

Comment:
The hint may be in the `with` clause, for example, `with cte as (select * from t use index(idx)) select * from cte`.

Reply:
For the hint in the with clause, we can handle them like:
`with cte as (select b+1 from t use index(expression_index_5)) select * from cte;` 
to `with cte as (select b+1 from t) select /*+ use_index(@sel_2 t expression_index_5) */ * from cte;`

Please make sure the that query block is correct.

Reply:
> For the hint in the with clause, we can handle them like:
> `with cte as (select b+1 from t use index(expression_index_5)) select * from cte;`
> to `with cte as (select b+1 from t) select /*+ use_index(@sel_2 t expression_index_5) */ * from cte;`
> 
> Please make sure the that query block is correct.

it's ok, PTAL @wjhuang2016